### PR TITLE
nixos: get system type from pkgs instance

### PR DESCRIPTION
**Pull request by Misterio77**
###### Description of changes

This avoids repetition when a flake `nixosSystem` caller is passing a `pkgs` instance to the function.

For example, when passing `pkgs = nixpkgs.legacyPackages.x86_64-linux` it should not be required to also pass `system = "x86_64-linux"`.

###### Examples

This change makes it possible to have similar nixosSystem and homeManagerConfiguration calls. For example:
```nix
{
  # ...
  outputs = { nixpkgs, home-manager, ... }: {

    nixosConfigurations = {
      "foo" = nixpkgs.lib.nixosSystem {
        pkgs = nixpkgs.legacyPackages.x86_64-linux;
        modules = [ ./nixos/configuration.nix ];
      };
    };

    homeConfigurations = {
      "bob@foo" = home-manager.lib.homeManagerConfiguration {
        pkgs = nixpkgs.legacyPackages.x86_64-linux;
        modules = [ ./home-manager/home.nix ];
      };
    };

  };
}
```

This also helps with more complex setups (passing `pkgs` is specially useful when using overlays):
```nix
{
  # ...
  outputs = { nixpkgs, home-manager, ... }:
  let
    forAllSystems = with nixpkgs.lib; genAttrs systems.flakeExposed;
  in rec {
    overlays.default = import ./overlay;

    legacyPackages = forAllSystems (system:
      import nixpkgs { inherit system; overlays = [ overlays.default ]; }
    );

    nixosConfigurations = {
      "foo" = nixpkgs.lib.nixosSystem {
        pkgs = legacyPackages.aarch64-linux;
        modules = [ ./hosts/foo ];
      };
      "bar" = nixpkgs.lib.nixosSystem {
        pkgs = legacyPackages.x86_64-linux;
        modules = [ ./hosts/bar ];
      };
    };

    homeConfigurations = {
      "bob@foo" = home-manager.lib.homeManagerConfiguration {
        pkgs = legacyPackages.aarch64-linux;
        modules = [ ./homes/bob ];
      };
      "alice@bar" = home-manager.lib.homeManagerConfiguration {
        pkgs = legacyPackages.x86_64-linux;
        modules = [ ./homes/alice ];
      };
    };
  };
}
```

`system` will still default to `builtins.currentSystem` (which, by design, does not work under flakes) when `pkgs` is not specified.
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
I'm making `system` parameter optional more generally here https://github.com/NixOS/nixpkgs/pull/177012, so that `nixos-generate-config`-based configs don't have to specify anything either. Another benefit is the consistency of doing everything within the module system rather than "before" it. Similarly, we could start to remove the `nixosSystem` `pkgs` parameter, so that it can also be specified as part of the config rather than as custom non-module logic.
tl;dr I think this logic should be moved to the `pkgsModule` after #177012.

We'll also want to set the new `nixpkgs.hostPlatform` and `nixpkgs.buildPlatform` options in that module.

**Comment by Misterio77** [comment]
Seems like a very nice improvement, thanks! Reducing the amount of arguments is very worthwile goal.

Quick question: in a future where `nixosSystem` no longer accepts `system` or `pkgs`, the right way to pass inputs overlays (or the evaluated `pkgs` instance itself) would be by passing it to the modules (that would then set `nixpkgs` options) through `specialArgs`?

It seems consuming overlays is a common usecase, so perhaps an optional `pkgs` argument on `nixosSystem` would still make sense?

**Comment by roberth** [comment]
> It seems consuming overlays is a common usecase, so perhaps an optional pkgs argument on nixosSystem would still make sense?

You can also consume overlays via [`nixpkgs.overlays`](https://search.nixos.org/options?channel=22.05&show=nixpkgs.overlays&from=0&size=50&sort=relevance&type=packages&query=nixpkgs.overlays), but `pkgs` is a better choice if you evaluate multiple machines at once.

> in a future where `nixosSystem` no longer accepts `pkgs`, the right way to pass overlays (or the evaluated `pkgs` instance itself) from the flake would be through `specialArgs`?

You could do it by adding a module. Something like this:

```
nixosSystem {
  modules = [
    ./configuration.nix
    {
      # With support for module-provided nixpkgs.overlays
      nixpkgs.pkgs = pkgs;

      # Alternative that rejects module-provided nixpkgs.overlays
      # Similar to specialArgs but not quite as absolute
      # _module.args.pkgs = lib.mkForce pkgs;
    }
  ];
}
```


`pkgs` in NixOS is more complicated than I think it should be. `specialArgs` and the above make for three significantly different ways of setting it. It's not easy to remove anything though. Probably best to document "setting pkgs" properly first.


> Or maybe somehow improving the UX for passing down args? A nixosSystem function that only takes modules and args looks really sweet.

I agree, though we should probably keep `specialArgs` because `args` is impossible to search for. Or perhaps you mean `_module.args`, but that one can't be used in `imports`.
Another idea is to add an optional `module` argument for a slightly more pleasant syntax. It does increase the number of ways we can do things though, so not sure.

**Comment by Misterio77** [comment]
Thanks for the clarifications!

Would this PR still be merged, at least until `pkgs` is deprecated?

With #177012, maybe `system` could also fallback to `null` when `builtins.currentSystem` is not available (as to avoid having to set `system = null` on the flake)?

Something like:
```nix
system ? pkgs.system or builtins.currentSystem or null
```

**Comment by roberth** [comment]
> Do you think it makes sense to still merge this PR, at least until `pkgs` is deprecated?

It would be better to set `nixpkgs.{host,build}Platform` instead, as those are more accurate and future proof than `system`, but that may be too disruptive for this release cycle, because we can't force users to migrate their configs to use these options yet, and they're mutually exclusive, to avoid ambiguities.

In theory, all modules should/can/do read the `pkgs` module argument (including `pkgs.stdenv.hostPlatform`) instead of reading any `nixpkgs.*` values, which means that we generally shouldn't have to set any of those `system`-like options when we bypass most of `nixos/.../nixpkgs.nix`.
Or arguments since #177012.
Is it not possible for you to remove the `system` argument from your flake's `nixosSystem` call after #177012? If not, what is the error message and trace?

> With #177012, maybe `system` could also fallback to `null` when `builtins.currentSystem` is not available (as to avoid having to set `system = null` on the flake)?

That's a sensible addition to help non-flake `pure-eval = true` users. `lib.nixosSystem` should force `system = null` though, even in impure mode, so I don't think we can remove it from the flake.

**Comment by Misterio77** [comment]
Oh, I totally missed the `system = null` change on the `nixosConfigurations` flake function!

Just bumped nixpkgs today and realised this totally solves what this PR was created for. Closing this.

Thanks again! The new system options seem a lot better for cross compiling entire nixos configurations.
